bson/primitive_codecs.go

```diff
@@ -45,15 +45,29 @@ func (pc PrimitiveCodecs) RegisterPrimitiveCodecs(rb *bsoncodec.RegistryBuilder)
 
 // RawValueEncodeValue is the ValueEncoderFunc for RawValue.
 //
-// Deprecated: Use bson.NewRegistry to get a registry with all primitive encoders and decoders
-// registered.
+// If the RawValue's Type is "invalid" and the RawValue's Value is not empty or
+// nil, then this method will return an error.
+//
+// Deprecated: Use bson.NewRegistry to get a registry with all primitive
+// encoders and decoders registered.
 func (PrimitiveCodecs) RawValueEncodeValue(_ bsoncodec.EncodeContext, vw bsonrw.ValueWriter, val reflect.Value) error {
 	if !val.IsValid() || val.Type() != tRawValue {
-		return bsoncodec.ValueEncoderError{Name: "RawValueEncodeValue", Types: []reflect.Type{tRawValue}, Received: val}
+		return bsoncodec.ValueEncoderError{
+			Name:     "RawValueEncodeValue",
+			Types:    []reflect.Type{tRawValue},
+			Received: val,
+		}
 	}
 
 	rawvalue := val.Interface().(RawValue)
 
+	if rawvalue.Type.String() == "invalid" {
+		return bsoncodec.ValueEncoderError{
+			Name:  "RawValueEncodeValue",
+			Types: []reflect.Type{tRawValue},
+		}
```

Comment:
This error value will result in an error message like:
```go
"RawValueEncodeValue can only encode valid reflect.Kind, but got "
```
which doesn't really describe the reason for the error. Instead, we should return an error that specifically describes the error condition. Something like:
```go
fmt.Errorf("the RawValue Type specifies an invalid BSON type: %#x", rawvalue.Type)
// E.g. error: "the RawValue Type specifies an invalid BSON type: 0x1e"
```

Reply:
The error will result in a message like this:

```
RawValueEncodeValue can only encode valid bson.RawValue, but got invalid
```

See [here](https://github.com/mongodb/mongo-go-driver/blob/master/bson/bsoncodec/bsoncodec.go#L83) for the logic.

Reply:
You're correct, I misunderstood the logic in `ValueEncoderError.Error` (and didn't test it 🤦 ). Nevertheless, I'm concerned that users won't know what to fix if they got that error message because it's not immediately obvious that the error is caused by `RawValue.Type` being invalid. We should try to provide an error message that gives users a clear action to resolve the error.

Reply:
Done!